File: private/services/edge/key-custodian/client-ts/src/sealing/keyring-backed-payload-sealer.ts
// -----------------------------------------------------------------------
// Copyright (c) DCSV. All rights reserved.
// -----------------------------------------------------------------------

import {
  type IPayloadSealer,
  PayloadSealer,
  type RecipientPublicKeyring,
} from "@dcsv-io/d2-encryption";
import type { ILogger } from "@dcsv-io/d2-logging";
import { RetryHelper } from "@dcsv-io/d2-resilience";
import { type D2Result, HttpStatusCode } from "@dcsv-io/d2-result";

import type { SealingClient } from "./sealing-client.js";

/** Default bounded refresh attempts before serving the current keyring. */
const _DEFAULT_REFRESH_ATTEMPTS = 3;

/** Default base backoff between bounded rotation-refresh attempts (ms). */
const _DEFAULT_REFRESH_BASE_DELAY_MS = 2_000;

/**
 * Transient-result classifier for the rotation-refresh retry — the TS twin of the
 * .NET `RetryHelper.RetryD2ResultAsync` default `ShouldRetry` (retries only
 * `ServiceUnavailable` / rate-limited, never a permanent auth/validation failure).
 */
const isTransientResult = (result: unknown): boolean => {
  const { statusCode } = result as D2Result<unknown>;

  return (
    statusCode === HttpStatusCode.ServiceUnavailable ||
    statusCode === HttpStatusCode.TooManyRequests
  );
};

/** Tunables for the hot-swap sealer. */
export interface KeyringBackedSealerOptions {
  readonly logger: ILogger;
  readonly refreshAttempts?: number;
  readonly refreshBaseDelayMs?: number;
}

/**
 * A KeyCustodian-backed {@link IPayloadSealer} with LAZY public-key fetch — the
 * TS twin of the .NET `KeyringBackedPayloadSealer`. The first `seal` triggers the
 * fetch (a producer host must not fail BOOT because a recipient has not lazily
 * provisioned yet); a failed fetch surfaces as a typed error the publisher maps
 * to a retryable failure, NEVER a plaintext fallback. Rotation re-fetches the
 * recipient's public keyring and swaps the reference (public material — no
 * zeroize; the swap still lets in-flight seals finish on a coherent keyring).
 */
export class KeyringBackedPayloadSealer implements IPayloadSealer {
  #sealer: PayloadSealer | undefined;
  // The in-flight first-fetch, shared by concurrent first-sealers so a cold-start
  // burst issues ONE fetch (the .NET twin dedupes under `r_initLock`). Cleared on
  // settle so a failed fetch re-attempts on the next seal.
  #initPromise: Promise<PayloadSealer> | undefined;
  #disposed = false;
  readonly #client: SealingClient;
  readonly #recipientServiceId: string;
  readonly #logger: ILogger;
  readonly #refreshAttempts: number;
  readonly #refreshBaseDelayMs: number;

  private constructor(
    client: SealingClient,
    recipientServiceId: string,
    options: KeyringBackedSealerOptions,
  ) {
    this.#client = client;
    this.#recipientServiceId = recipientServiceId;
    this.#logger = options.logger;
    this.#refreshAttempts =
      options.refreshAttempts ?? _DEFAULT_REFRESH_ATTEMPTS;
    this.#refreshBaseDelayMs =
      options.refreshBaseDelayMs ?? _DEFAULT_REFRESH_BASE_DELAY_MS;
  }

  /**
   * Builds a lazy sealer — NOT async (no boot fetch; the first `seal` fetches).
   *
   * @param client The sealing client.
   * @param recipientServiceId The recipient service to seal to.
   * @param options Logger + tunables.
   */
  static create(
    client: SealingClient,
    recipientServiceId: string,
    options: KeyringBackedSealerOptions,
  ): KeyringBackedPayloadSealer {
    return new KeyringBackedPayloadSealer(client, recipientServiceId, options);
  }

  /** @inheritdoc */
  async seal(plaintext: Uint8Array): Promise<Uint8Array> {
    this.#throwIfDisposed();

    const sealer = this.#sealer ?? (await this.#ensureSealer());

    return sealer.seal(plaintext);
  }

  /**
   * Re-fetches the recipient public keyring and hot-swaps it. Bounded retry; on
   * exhaustion the current keyring is served (never a gap) with a loud log.
   */
  async refresh(): Promise<void> {
    this.#throwIfDisposed();

    // Bounded, transient-classified retry via the shared RetryHelper (the .NET
    // twin drives the identical refresh through `RetryHelper.RetryD2ResultAsync`):
    // a permanent auth/validation failure short-circuits immediately; a transient
    // failure backs off exponentially up to the cap, then serves the current keyring.
    const result = await RetryHelper.retryD2ResultAsync<RecipientPublicKeyring>(
      () => this.#client.getPublicKeyring(this.#recipientServiceId),
      {
        maxAttempts: this.#refreshAttempts,
        baseDelayMs: this.#refreshBaseDelayMs,
        shouldRetry: isTransientResult,
      },
    );

    if (!result.failed && result.data !== undefined) {
      this.#sealer = new PayloadSealer(result.data);

      return;
    }

    this.#logger.error(
      "sealed public keyring refresh exhausted — serving current keyring",
      {
        attempts: this.#refreshAttempts,
        recipientServiceId: this.#recipientServiceId,
        errorCode: result.errorCode,
      },
    );
  }

  /** Disposes the sealer. Public material carries no key bytes to zeroize. */
  dispose(): void {
    this.#disposed = true;
    this.#sealer = undefined;
    this.#initPromise = undefined;
  }

  // Dedupes concurrent first-seals onto one in-flight fetch, then installs the
  // sealer — unless a rotation refresh already installed a newer keyring while the
  // fetch was in flight (never overwrite a fresher keyring with a staler one).
  #ensureSealer(): Promise<PayloadSealer> {
    this.#initPromise ??= this.#initializeSealer();

    return this.#initPromise;
  }

  async #initializeSealer(): Promise<PayloadSealer> {
    try {
      const keyring = await this.#fetchKeyring();

      this.#sealer ??= new PayloadSealer(keyring);

      return this.#sealer;
    } finally {
      this.#initPromise = undefined;
    }
  }

  async #fetchKeyring(): Promise<RecipientPublicKeyring> {
    const result = await this.#client.getPublicKeyring(
      this.#recipientServiceId,
    );

    if (result.failed || result.data === undefined) {
      throw new Error(
        `KeyCustodian sealed public keyring fetch failed for recipient ` +
          `'${this.#recipientServiceId}' — cannot seal (retryable; never a ` +
          "plaintext fallback).",
      );
    }

    return result.data;
  }

  #throwIfDisposed(): void {
    if (this.#disposed) {
      throw new Error("KeyringBackedPayloadSealer has been disposed.");
    }
  }
}

File: private/services/edge/key-custodian/client/Sealing/KeyringBackedPayloadSealer.cs
// -----------------------------------------------------------------------
// <copyright file="KeyringBackedPayloadSealer.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace DcsvIo.D2.Private.Edge.KeyCustodian.Client.Sealing;

using DcsvIo.D2.Encryption;
using DcsvIo.D2.Private.Edge.KeyCustodian.Client.Keyring;
using DcsvIo.D2.Resilience.Retry;
using Microsoft.Extensions.Logging;

/// <summary>
/// A sealed, hot-swappable <see cref="IPayloadSealer"/> capability backed by a recipient
/// service's KeyCustodian PUBLIC seal keyring — the producer twin of
/// <see cref="KeyringBackedPayloadOpener"/>. Holds only wire-public key material and refreshes
/// atomically on a rotation event.
/// </summary>
/// <remarks>
/// <para>
/// <strong>Least-privilege surface.</strong> The public members are <see cref="Seal"/> /
/// <see cref="DisposeAsync"/> only. There is NO <c>Open</c> — a producer can never open any
/// sealed frame, including its own (the capability split is compile-time).
/// </para>
/// <para>
/// <strong>Lazy public-key fetch (deliberately asymmetric to the opener).</strong> A producer
/// host must not fail BOOT because a recipient has not lazily provisioned its sealing key yet,
/// so the first <see cref="Seal"/> triggers the fetch (bounded retry); a failed fetch surfaces
/// as a thrown exception the publisher maps to a retryable publish failure — NEVER a plaintext
/// fallback. Subsequent seals serve the cached keyring.
/// </para>
/// <para>
/// <strong>No zeroize.</strong> A <see cref="RecipientPublicKeyring"/> holds wire-public
/// material, so a rotation simply swaps the holder reference (in-flight seals finish on the
/// prior coherent keyring; GC reclaims it) — no grace-dispose / zeroize machinery is needed.
/// </para>
/// </remarks>
public sealed class KeyringBackedPayloadSealer : IPayloadSealer, IAsyncDisposable
{
    private const int _REFRESH_MAX_ATTEMPTS = 3;
    private static readonly TimeSpan sr_refreshBackoffBase = TimeSpan.FromSeconds(2);

    // The lazy first fetch is bounded so a connected-but-unresponsive KeyCustodian cannot hang
    // a publish thread indefinitely (fail-loud → retryable publish failure on expiry).
    private static readonly TimeSpan sr_lazyFetchTimeout = TimeSpan.FromSeconds(30);

    private readonly string r_sealDomain;
    private readonly string r_recipientServiceId;
    private readonly ISealingClient r_client;
    private readonly IRotationEventChannel r_channel;
    private readonly ILogger<KeyringBackedPayloadSealer> r_logger;
    private readonly RetryOptions<D2Result<RecipientPublicKeyring>> r_refreshRetryOptions;
    private readonly Lock r_initLock = new();
    private readonly TimeSpan r_lazyFetchTimeout;

    private Holder? _holder;
    private IAsyncDisposable? _subscription;
    private int _disposed;

    private KeyringBackedPayloadSealer(
        string recipientServiceId,
        ISealingClient client,
        IRotationEventChannel channel,
        ILogger<KeyringBackedPayloadSealer> logger,
        int maxRefreshAttempts,
        TimeSpan refreshBackoffBase,
        TimeSpan lazyFetchTimeout)
    {
        r_recipientServiceId = recipientServiceId;
        r_sealDomain = SealDomainName.For(recipientServiceId);
        r_client = client;
        r_channel = channel;
        r_logger = logger;
        r_lazyFetchTimeout = lazyFetchTimeout;
        r_refreshRetryOptions = new RetryOptions<D2Result<RecipientPublicKeyring>>
        {
            MaxAttempts = maxRefreshAttempts,
            BaseDelayMs = (int)refreshBackoffBase.TotalMilliseconds,
        };
    }

    /// <inheritdoc />
    public byte[] Seal(ReadOnlySpan<byte> plaintext)
    {
        var holder = Volatile.Read(ref _holder) ?? EnsureInitialized();
        return holder.Sealer.Seal(plaintext);
    }

    /// <inheritdoc />
    public async ValueTask DisposeAsync()
    {
        if (Interlocked.Exchange(ref _disposed, 1) == 1)
            return;

        // Snapshot + clear the subscription under the SAME lock a first-seal subscribes
        // under, so an initialization racing this dispose can never leave a subscription
        // registered after this snapshot: the subscribe and this read are mutually
        // exclusive, so this always observes a subscription the init installed (no leak).
        IAsyncDisposable? subscription;

        lock (r_initLock)
        {
            subscription = _subscription;
            _subscription = null;
        }

        if (subscription is not null)
            await subscription.DisposeAsync().ConfigureAwait(false);
    }

    /// <summary>Returns the type name only — never any keyring contents.</summary>
    public override string ToString() => nameof(KeyringBackedPayloadSealer);

    /// <summary>
    /// Builds a <see cref="KeyringBackedPayloadSealer"/> for
    /// <paramref name="recipientServiceId"/>. Does NOT fetch — the first <see cref="Seal"/>
    /// lazily fetches the recipient's public keyring (so a producer host never fails boot on a
    /// not-yet-provisioned recipient).
    /// </summary>
    /// <param name="recipientServiceId">The recipient service to seal payloads to.</param>
    /// <param name="client">The (internal) fetch seam.</param>
    /// <param name="channel">The rotation-notification channel (subscribed on first seal).</param>
    /// <param name="logger">Logger.</param>
    /// <returns>A ready-to-serve sealer capability (fetch deferred to first seal).</returns>
    internal static KeyringBackedPayloadSealer Create(
        string recipientServiceId,
        ISealingClient client,
        IRotationEventChannel channel,
        ILogger<KeyringBackedPayloadSealer> logger)
        => Create(
            recipientServiceId,
            client,
            channel,
            logger,
            _REFRESH_MAX_ATTEMPTS,
            sr_refreshBackoffBase,
            sr_lazyFetchTimeout);

    /// <summary>
    /// Test-only <c>Create</c> overload injecting the rotation-refresh retry budget + the lazy
    /// fetch timeout so the retry / timeout behavior can be exercised deterministically.
    /// </summary>
    /// <param name="recipientServiceId">The recipient service to seal payloads to.</param>
    /// <param name="client">The fetch seam.</param>
    /// <param name="channel">The rotation channel.</param>
    /// <param name="logger">Logger.</param>
    /// <param name="maxRefreshAttempts">Bounded rotation-refresh attempt cap.</param>
    /// <param name="refreshBackoffBase">Per-attempt backoff base.</param>
    /// <param name="lazyFetchTimeout">The bounded lazy first-fetch timeout.</param>
    /// <returns>A ready-to-serve sealer with the injected timings.</returns>
    internal static KeyringBackedPayloadSealer CreateForTesting(
        string recipientServiceId,
        ISealingClient client,
        IRotationEventChannel channel,
        ILogger<KeyringBackedPayloadSealer> logger,
        int maxRefreshAttempts,
        TimeSpan refreshBackoffBase,
        TimeSpan lazyFetchTimeout)
        => Create(
            recipientServiceId,
            client,
            channel,
            logger,
            maxRefreshAttempts,
            refreshBackoffBase,
            lazyFetchTimeout);

    private static KeyringBackedPayloadSealer Create(
        string recipientServiceId,
        ISealingClient client,
        IRotationEventChannel channel,
        ILogger<KeyringBackedPayloadSealer> logger,
        int maxRefreshAttempts,
        TimeSpan refreshBackoffBase,
        TimeSpan lazyFetchTimeout)
    {
        recipientServiceId.ThrowIfFalsey();
        ArgumentNullException.ThrowIfNull(client);
        ArgumentNullException.ThrowIfNull(channel);
        ArgumentNullException.ThrowIfNull(logger);

        return new KeyringBackedPayloadSealer(
            recipientServiceId,
            client,
            channel,
            logger,
            maxRefreshAttempts,
            refreshBackoffBase,
            lazyFetchTimeout);
    }

    // Double-checked lazy init: the first seal fetches the public keyring (bounded), installs
    // the holder, and subscribes to rotation. Concurrent first-sealers serialize on the lock
    // and observe the single installed holder. A failed fetch throws (retryable publish
    // failure) and leaves _holder null so the next seal re-attempts.
    private Holder EnsureInitialized()
    {
        ObjectDisposedException.ThrowIf(Volatile.Read(ref _disposed) == 1, this);

        lock (r_initLock)
        {
            var existing = Volatile.Read(ref _holder);

            if (existing is not null)
                return existing;

            using var fetchCts = new CancellationTokenSource(r_lazyFetchTimeout);
            var fetch = r_client.GetPublicKeyringAsync(r_recipientServiceId, fetchCts.Token)
                .AsTask().GetAwaiter().GetResult();

            if (!fetch.CheckSuccess(out var keyring) || keyring is null)
            {
                var errorCode = fetch.ErrorCode ?? SealingMetrics.Tags.NONE;
                SealingMetrics.RecordFetch(r_sealDomain, success: false);
                SealingLog.SealSealerLazyFetchFailed(r_logger, r_sealDomain, errorCode);

                throw new InvalidOperationException(
                    $"Seal public-keyring fetch for domain '{r_sealDomain}' failed (error "
                    + $"'{errorCode}'); the payload cannot be sealed. This is a retryable "
                    + "publish failure — the payload is NEVER shipped as plaintext.");
            }

            var holder = new Holder(new PayloadSealer(keyring));

            // A DisposeAsync that set _disposed during the (bounded) fetch above is now
            // blocked on r_initLock behind this init; there is no point subscribing if
            // disposal has begun — fail loud instead of registering a doomed subscription.
            ObjectDisposedException.ThrowIf(Volatile.Read(ref _disposed) == 1, this);

            // Subscribe once, on first successful init. The subscribe + the _subscription
            // publish happen under the SAME lock DisposeAsync snapshots under, so a racing
            // dispose always observes (and disposes) this subscription — no leak. A rotation
            // racing this init is fine: the holder is installed before the swap can occur.
            _subscription = r_channel.Subscribe(r_sealDomain, OnRotationAsync);
            Volatile.Write(ref _holder, holder);

            return holder;
        }
    }

    private async Task OnRotationAsync(CancellationToken ct)
    {
        var fetch = await RetryHelper.RetryD2ResultAsync(
            (_, token) => r_client.GetPublicKeyringAsync(r_recipientServiceId, token),
            r_refreshRetryOptions,
            ct).ConfigureAwait(false);

        if (fetch.CheckSuccess(out var keyring) && keyring is not null)
        {
            // Public material: swap the holder reference atomically (no zeroize / grace).
            // In-flight seals finish on the prior coherent keyring; GC reclaims it.
            Volatile.Write(ref _holder, new Holder(new PayloadSealer(keyring)));
            SealingMetrics.RecordRotationHotSwap(r_sealDomain);
            SealingLog.SealKeyringRefreshSucceeded(r_logger, r_sealDomain, keyring.ActiveKid);

            return;
        }

        // Terminal: keep serving the current keyring (a later rotation / restart re-drives).
        var errorCode = fetch.ErrorCode ?? SealingMetrics.Tags.NONE;
        SealingMetrics.RecordRefreshFailure(r_sealDomain, errorCode);
        SealingLog.SealKeyringRefreshFailed(r_logger, r_sealDomain, errorCode);
    }

    // Only the sealer is held — a RecipientPublicKeyring is wire-public (no zeroize / dispose),
    // so unlike the opener's holder the keyring reference is not retained past construction.
    private sealed record Holder(IPayloadSealer Sealer);
}
